h2/src/main/org/h2/mvstore/db/ValueDataType.java

```diff
@@ -136,7 +156,12 @@ private int compareValues(Value a, Value b, int sortType) {
         if (aNull || bNull) {
             return SortOrder.compareNull(aNull, sortType);
         }
-        int comp = a.compareTypeSafe(b, compareMode);
+
+        int t2 = Value.getHigherOrder(a.getType(), b.getType());
+        String[] enumerators = ValueEnum.getEnumeratorsForBinaryOperation(a, b);
```

Comment:
This method allocates a new empty array on each invocation if types are not enumerations. It's cheap enough, but actually it's not designed to be used if higher order is not `Value.ENUM`. The following code is used everywhere:
```Java
int comp;
if (t2 == Value.ENUM) {
    String[] enumerators = ValueEnum.getEnumeratorsForBinaryOperation(a, b);
    a = a.convertToEnum(enumerators);
    b = b.convertToEnum(enumerators);
} else {
    a = a.conventTo(t2, -1, mode)
    b = b.convertTo(t2, -1, mode);
}
```
(May be we need a specialized `convertTo()` with an additional argument with the second operand to use enumerators from it if type is not `ENUM` and target type is `ENUM` to avoid all these ugly patterns.)